crates/shared/src/price_estimation/quasimodo.rs

```diff
@@ -0,0 +1,346 @@
+use crate::bad_token::BadTokenDetecting;
+use crate::baseline_solver::BaseTokens;
+use crate::http_solver_api::model::{
+    AmmModel, AmmParameters, BatchAuctionModel, ConstantProductPoolParameters, CostModel, FeeModel,
+    OrderModel, TokenInfoModel,
+};
+use crate::http_solver_api::HttpSolverApi;
+use crate::price_estimation::gas::{
+    ERC20_TRANSFER, GAS_PER_ORDER, GAS_PER_UNISWAP, INITIALIZATION_COST, SETTLEMENT,
+};
+use crate::price_estimation::{
+    ensure_token_supported, Estimate, PriceEstimating, PriceEstimationError, Query,
+};
+use crate::recent_block_cache::Block;
+use crate::sources::uniswap_v2::pool_cache::PoolCache;
+use crate::sources::uniswap_v2::pool_fetching::PoolFetching;
+use crate::token_info::TokenInfoFetching;
+use ethcontract::{H160, U256};
+use gas_estimation::GasPriceEstimating;
+use model::order::OrderKind;
+use model::TokenPair;
+use num::{BigInt, BigRational};
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+pub struct QuasimodoPriceEstimator {
+    pub api: Arc<HttpSolverApi>,
+    pub pools: Arc<PoolCache>,
+    pub bad_token_detector: Arc<dyn BadTokenDetecting>,
+    pub token_info: Arc<dyn TokenInfoFetching>,
+    pub gas_info: Arc<dyn GasPriceEstimating>,
+    pub native_token: H160,
+    pub base_tokens: Arc<BaseTokens>,
+}
+
+impl QuasimodoPriceEstimator {
+    async fn estimate(&self, query: &Query) -> Result<Estimate, PriceEstimationError> {
+        if query.buy_token == query.sell_token {
+            return Ok(Estimate {
+                out_amount: query.in_amount,
+                gas: 0.into(),
+            });
+        }
+
+        ensure_token_supported(query.buy_token, self.bad_token_detector.as_ref()).await?;
+        ensure_token_supported(query.sell_token, self.bad_token_detector.as_ref()).await?;
+
+        let gas_cost = U256::from_f64_lossy(self.gas_info.estimate().await?.effective_gas_price());
+
+        let mut tokens = self.base_tokens.tokens().clone();
+        tokens.insert(query.sell_token);
+        tokens.insert(query.buy_token);
+        tokens.insert(self.native_token);
+        let tokens: Vec<_> = tokens.drain().collect();
+
+        let token_infos = self.token_info.get_token_infos(&tokens).await;
+
+        let tokens = tokens
+            .iter()
+            .map(|token| {
+                (
+                    *token,
+                    TokenInfoModel {
+                        decimals: token_infos[token].decimals,
```

Comment:
Can't do this because get_token_infos will not have entries in hash map on errors.

Reply:
Well we always return an entry: https://github.com/gnosis/gp-v2-services/blob/153217e3a3c6d0ca3967eef0b4a0e6cddb2d8489/crates/shared/src/token_info.rs#L56-L65

But you're right, we shouldn't count on implementation details.

Reply:
I didn't realize it would just set them to None, I see. In that case I would also be fine with changing the doc comment to mention this.